expression/integration_test/integration_test.go

```diff
@@ -3782,13 +3782,10 @@ func TestShardIndexOnTiFlash(t *testing.T) {
 	}
 	tk.MustExec("set @@session.tidb_enforce_mpp = 0")
 	tk.MustExec("set @@session.tidb_allow_mpp = 0")
-	rows = tk.MustQuery("explain select max(b) from t").Rows()
-	for _, row := range rows {
-		line := fmt.Sprintf("%v", row)
-		if strings.Contains(line, "TableFullScan") {
-			require.NotContains(t, line, "mpp[tiflash]")
-		}
-	}
+	// when we isolated the read engine as 'tiflash' and banned TiDB opening allow-mpp, no suitable plan is generated.
+	_, err := tk.Exec("explain select max(b) from t")
+	require.NotNil(t, err)
+	require.Equal(t, err.Error(), "[planner:1815]Internal : Can't find a proper physical plan for this query")
```

Comment:
I remember err msg has already refined(https://github.com/pingcap/tidb/pull/43908/files#diff-bca2046fd17f493604dd1124ab98d60de4aa21e57135324e365eb33e68b74f43), why not works here?

Reply:
make sense, I will check allowTiflashCop-Switch there like you do

Reply:
as we talked about, t.invalid() is a lazy check, we couldn't tell what exactly cause the plan pruning and finally nothing generated